### PR TITLE
add path parameter to .include macro

### PR DESCRIPTION
**Pull request by allanjude**
Allow searching an array of paths for includes

include macro gains new parameter 'paths', an array of paths to search
if glob=false (default), first match is included
if glob=true all matches are included

New API method:
bool ucl_set_include_path(parser, ucl_object_array)

excepts a UCL_ARRAY of paths to search for the include filename

Usage:
ucl_object_t *obj;

obj = ucl_object_typed_new(UCL_ARRAY);
ucl_array_append(obj, ucl_object_fromstring("/etc/bhyve.d"));
ucl_array_append(obj, ucl_object_fromstring("/usr/local/etc/bhyve.d"));

ucl_set_include_path(parser, obj);
ucl_object_free(obj);

### CODE REVIEW

File: src/ucl_util.c [hunk 1]
```diff
@@ -1131,15 +1154,48 @@ ucl_include_common (const unsigned char *data, size_t len,
 		}
 	}
 
-	if (*data == '/' || *data == '.') {
-		/* Try to load a file */
-		return ucl_include_file (data, len, parser, need_sign, !try_load,
-				allow_glob, priority);
+	if (parser->includepaths == NULL) {
+		if (allow_url && strnstr(data, "://", len) != NULL) {
```
Comment (vstakhov) [inline src/ucl_util.c]: `strnstr` is not standard function (it's FreeBSD specific), and hence it should not be used.

File: src/ucl_parser.c [hunk 2]
```diff
@@ -2300,3 +2301,25 @@ ucl_parser_add_string (struct ucl_parser *parser, const char *data,
...
+
+bool
+ucl_set_include_path (struct ucl_parser *parser, ucl_object_t *paths)
+{
+	if (parser == NULL || paths == NULL) {
+		return false;
+	}
+
+	if (parser->includepaths == NULL) {
+		parser->includepaths = ucl_object_copy(paths);
```
Comment (vstakhov) [inline src/ucl_parser.c]: Space is missing before opening branch (and in the subsequent places as well).

File: src/ucl_util.c [hunk 3]
```diff
@@ -1123,6 +1141,11 @@ ucl_include_common (const unsigned char *data, size_t len,
 					allow_url =  ucl_object_toboolean (param);
 				}
 			}
+			else if (param->type == UCL_ARRAY) {
+				if (strcmp (param->key, "path") == 0) {
```
Comment (vstakhov) [inline src/ucl_util.c]: Key might be not null terminated. You should use `keylen` here.

File: src/ucl_util.c [hunk 4]
```diff
@@ -1123,6 +1141,11 @@ ucl_include_common (const unsigned char *data, size_t len,
...
+					ucl_set_include_path(parser, param);
```
Comment (vstakhov) [inline src/ucl_util.c]: Space.

File: src/ucl_util.c [hunk 5]
```diff
@@ -1131,15 +1154,48 @@ ucl_include_common (const unsigned char *data, size_t len,
...
+			return ucl_include_file (data, len, parser, need_sign, !try_load,
+					allow_glob, false, priority);
+		}
 	}
-	else if (allow_url) {
-		/* Globbing is not used for URL's */
-		return ucl_include_url (data, len, parser, need_sign, !try_load,
-				priority);
+	else {
+		if (allow_url && strnstr(data, "://", len) != NULL) {
```
Comment (vstakhov) [inline src/ucl_util.c]: `strnstr`

File: src/ucl_util.c (hunk 3 above)
Comment (allanjude) [inline src/ucl_util.c]: None of the other cases in this if ladder use keylen, should they all be changed to strncmp?

**Pull request merged**